### DISCUSSION TITLE
publish stream on cellular network

### DISCUSSION THREAD

**Discussion by gadididi** [bluenviron/mediamtx#296] [Questions]
## Which version are you using?

gst-launch-1.0
Linux 18.04

## **Question**

I publish stream from the drone (DJI) to the RTSP server by building a pipeline of rtspclientsink. ( i guess the RTSP server share the stream with RTP)
On the localhost on my DJI computer(without using a cellular network), it was played very well, but when I publish the stream with the cellular network the client-side can connect to the RTSP server, but the frames stuck...
I use (appsrc h264parse video/x-h264 rtspclientsink location=rtsp://localhost:8554/mystream)

* I can saw in the Wireshark that there are a traffic of RTP/UDP packets between the drone computer and the cellphone(client) 

**my question is: is there a way with this server to get a publish stream, and scale it or make some manipulation on it (maybe in the ymal file?)**

**second question** 
* I build a little rtsp server that streams only to GO RTSP server (this repository)  and this server stream it to everyone.
the command line stream _"appsrc name=mysrc  ! queue ! decodebin ! videorate ! video/x-raw,framerate=24/1 !x264enc speed-preset=superfast tune=zerolatency ! h264parse ! rtph264pay name=pay0 pt=96 "_

my question is: is there a way in this GO rtsp server to recognize if the bandwidth low and to reduce the resulotion  quality 

thank you all :)
Gadi

**Comment by aler9** [comment]
Hello,

> is there a way with this server to get a publish stream, and scale it or make some manipulation on it 

Yes, use ffmpeg together with rtsp-simple-server to manipulate the video, and publish the result on another path, like explained in the readme:

https://github.com/aler9/rtsp-simple-server#remuxing-re-encoding-compression

For instance, if your video is published on `/mystream`, you can pull the video from this path, change the resolution and publish it on `/mystream2`, with this configuration

```yml
paths:
  all:
  mystream:
    runOnPublish: ffmpeg -i rtsp://localhost:$RTSP_PORT/$RTSP_PATH -vf scale=320:240 -c:v libx264 -f rtsp rtsp://localhost:$RTSP_PORT/mystream2
    runOnPublishRestart: yes
```

> my question is: is there a way in this GO rtsp server to recognize if the bandwidth low and to reduce the resulotion quality

<s>Unfortunately this is not possible with the RTSP protocol - the only protocol i know that is capable of doing this is WebRTC, but it must be supported natively by the DJI drone (and is not).</s>

The RTSP protocol allows to implement this feature, since the server sends periodically to a publisher RTCP packets which contain QOS informations, that allows to tune the bandwidth, but the source (the DJI drone) must support the feature natively, and it doesn't.

**Comment by gadididi** [comment]
#

> Hello,
> 
> > is there a way with this server to get a publish stream, and scale it or make some manipulation on it
> 
> Yes, use ffmpeg together with rtsp-simple-server to manipulate the video, and publish the result on another path, like explained in the readme:
> 
> https://github.com/aler9/rtsp-simple-server#remuxing-re-encoding-compression
> 
> For instance, if your video is published on `/mystream`, you can pull the video from this path, change the resolution and publish it on `/mystream2`, with this configuration
> 
> ```yaml
> paths:
>   all:
>   mystream:
>     runOnPublish: ffmpeg -i rtsp://localhost:$RTSP_PORT/$RTSP_PATH -vf scale=320:240 -c:v libx264 -f rtsp rtsp://localhost:$RTSP_PORT/mystream2
>     runOnPublishRestart: yes
> ```
> 
> > my question is: is there a way in this GO rtsp server to recognize if the bandwidth low and to reduce the resulotion quality
> 
> Unfortunately this is not possible with the RTSP protocol - the only protocol i know that is capable of doing this is WebRTC, but it must be supported natively by the DJI drone (and is not).

Hi, @aler9 thank you a lot! 
I try to re-encoded by RTSP simple server with GStreamer 
my line is: 
_gst-launch-1.0 rtspsrc location=rtsp://localhost:$RTSP_PORT/$RTSP_PATH ! queue ! decodebin ! videoscale ! videorate ! video/x-raw,framerate=24/1,width=480,height=320 ! x264enc  speed-preset=ultrafast tune=zerolatency ! h264parse ! rtph264pay name=pay0 pt=96_
but i need rtsp sink , do i need to use rtspclientsink ? 

i run for example :
**gst-launch-1.0 rtspsrc location=rtsp://localhost:8554/mystream ! queue ! decodebin ! videoscale ! videorate ! video/x-raw,framerate=24/1,width=480,height=320 ! x264enc bitrate= 500 speed-preset=ultrafast tune=zerolatency ! h264parse ! rtspclientsink location=rtsp://localhost:8554/mystream2**

and I try to connect from the GStreamer client and I get on the RTSP simple server-side:

_ERR: no one is publishing to path 'mystream2'_

again thank you!

**Comment by aler9** [comment]
> gst-launch-1.0 rtspsrc location=rtsp://localhost:8554/mystream ! queue ! decodebin ! videoscale ! videorate ! video/x-raw,framerate=24/1,width=480,height=320 ! x264enc bitrate= 500 speed-preset=ultrafast tune=zerolatency ! h264parse ! rtspclientsink location=rtsp://localhost:8554/mystream2

i tested this exact pipeline, without touching anything, and it works like a charm - make sure that it is publishing to the server, and only after, try to connect a client.

[a comment by an automated account is not shown]
